src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java

```diff
@@ -83,9 +79,27 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
 
     @Override
     public Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        MatchAllQueryBuilder query = new MatchAllQueryBuilder();
+        query.fromXContent(parseContext);
+        return query.toQuery(parseContext);
```

Comment:
I guess we could generify this on the base class since the last two lines are going to be the same everywhere? and have a new abstract method `newQuery()` that creates a new instance of the query(Builder) ?